### PR TITLE
Add combinational components

### PR DESCRIPTION
**Pull request by andrewb1999**
Solves #1054 

Added some simple tests and updated papercut to make sure the combinational component only contains other combinational cells. Let me know if there are more changes that should be made before merging.

### CODE REVIEW

**Comment by rachitnigam** [comment]
Argh, seems like github actions continues to be annoying about external contributions and not running actions. @andrewb1999 can I give you write access to the repo so you can place the code on a branch and have GH run the tests?

**Comment by andrewb1999** [comment]
Yes, that works.

**Comment by rachitnigam** [comment]
We should also disallow `group` or `comb group` definitions within a `comb component` since there is no way to schedule their execution

**Comment by andrewb1999** [comment]
Good point. I will update.

**Comment by sampsyo** [comment]
Wow; this is super awesome! Thanks for jumping on this, @andrewb1999!!!!

**Comment by andrewb1999** [comment]
I added a check in papercut on my local branch before realizing that we can't even get to that point before getting an error. Groups error out for not assigning to a done and comb groups error out for not being used. Is there some place earlier in the flow I should be checking for this?

**Comment by rachitnigam** [comment]
Yeah, those checks should go in the well-formed pass. Papercut is for things that _may_ be errors but well-formed is to ensures that the program is well-defined. If well-formed fails, there are no guarantees

**Comment by andrewb1999** [comment]
Makes sense, moved those checks to well-formed.

**Comment by rachitnigam** [comment]
Great, let me know when it's ready for review! Seems like GH actions decided to work after all

**Comment by andrewb1999** [comment]
Should be ready to review, but I'm not sure why CI is failing. Looks like there are errors in the frontend, can you take a look when you have a chance?

File: calyx/src/ir/printer.rs [hunk 1]
```diff
@@ -160,14 +160,25 @@ impl Printer {
...
 
-        writeln!(
-            f,
-            "component {}{}({}) -> ({}) {{",
-            comp.name.id,
-            Self::format_attributes(&comp.attributes),
-            Self::format_ports(&inputs),
-            Self::format_ports(&outputs),
-        )?;
+        if comp.is_comb {
```
Comment (rachitnigam) [inline calyx/src/ir/printer.rs:163]: Possible to simplify this by having a `let pre = if comp.is_come { "comb " } else { "" };`

File: calyx/src/passes/well_formed.rs [hunk 2]
```diff
@@ -185,6 +185,34 @@ impl Visitor for WellFormed {
             }
         }
 
+        // If the component is combinational, make sure all cells are also combinational
+        // and there are no group or comb group definitions
```
Comment (rachitnigam) [inline calyx/src/passes/well_formed.rs:189]: We also need to make sure that the control program is empty

File: tests/errors/comb-component-groups.expect [hunk 3]
```diff
@@ -0,0 +1,4 @@
+---CODE---
+1
+---STDERR---
+Error: Malformed Structure: Component `custom_lt` is marked combinational but contains a group.
```
Comment (rachitnigam) [inline tests/errors/comb-component-groups.expect:4]: It should be possible to print the location of the offending group here

File: calyx/src/passes/well_formed.rs (hunk 2 above)
Comment (andrewb1999) [reply calyx/src/passes/well_formed.rs:189]: Added a check for this, but there should be no way to create this based on the changes to the parser.

**Comment by andrewb1999** [comment]
Addressed comments.

**Comment by andrewb1999** [comment]
Seems like CI is stuck?

**Review by rachitnigam — APPROVED** [review]
LGTM!

**Comment by rachitnigam** [comment]
Sigh, yeah not sure why...

**Comment by rachitnigam** [comment]
Ugh, we seem to have annoyed the GH god I guess? @andrewb1999 can you create a branch in this repo and open a PR against it? Sorry about the hassle I've tried to change all the settings GH people have told me about but it doesn't work

**Comment by andrewb1999** [comment]
Closing and trying again.

**Pull request closed without merging**